modules/javafx.graphics/src/main/java/com/sun/javafx/scene/layout/SpaceDistributor.java

```diff
@@ -0,0 +1,258 @@
+        if (grow == (targetSpaceRequired < spaceRequired)) {
+            return quantizedSizes;
+        }
+
+        // if the target sizes are smaller/larger than the intended space, then return the target sizes:
+        if (grow == (targetSpaceRequired < quantizedSpace)) {
+            return quantizedTargetSizes;
+        }
+
+        distributeAvailablePreScaled(available, renderScale, quantizedSizes, quantizedTargetSizes);
+
+        return quantizedSizes;
+    }
+
+    /*
+     * This method modifies the given spaces array to the correct sizes given the available space
+     * and taken into account their maximum sizes. The input targetSizes array is not modified.
+     *
+     * Note: when renderScale is not zero, all input values must be on whole pixel boundaries.
+     * Furthermore, all newly calculated values in this method are also rounded to pixel boundaries.
+     */
+    private static void distributeAvailablePreScaled(double available, double renderScale, double[] sizes, double[] targetSizes) {
+        int count = sizes.length;
+        int resizables = 0;
+        boolean grow = available >= 0;
+
+        /*
+         * Determine the amount of spaces that can be resized, while
+         * checking the input values make sense:
+         */
+
+        for (int i = 0; i < count; i++) {
+            if (grow && sizes[i] > targetSizes[i]) {
+                throw new IllegalArgumentException("growing: sizes[" + i + "] exceeded targetSizes[" + i + "]: " + sizes[i] + " > " + targetSizes[i]);
+            }
+            else if(!grow && targetSizes[i] > sizes[i]) {
+                throw new IllegalArgumentException("shrinking: targetSizes[" + i + "] exceeded sizes[" + i + "]: " + targetSizes[i] + " > " + sizes[i]);
+            }
+
+            if (sizes[i] != targetSizes[i]) {
+                resizables++;
+            }
+        }
+
+        /*
+         * Distribute available space evenly among resizable spaces:
+         */
+
+        for (;;) {
+            double idealChange = down(available / resizables, renderScale);
+
+            if (idealChange == 0) {  // exit loop when nothing is going to change
+                break;
+            }
+
+            int initialResizables = resizables;
+
+            for (int i = 0; i < count; i++) {
+                double size = sizes[i];
+                double targetSize = targetSizes[i];
+
+                if ((grow && size < targetSize) || (!grow && size > targetSize)) {
+                    double x = round(size + idealChange, renderScale);
+                    double newSpace = grow ? Math.min(targetSize, x) : Math.max(targetSize, x);
+
+                    available = round(available - newSpace + size, renderScale);
+                    sizes[i] = newSpace;
+
+                    if (newSpace == targetSize && --resizables == 0) {  // exit method when nothing left to resize
```

Comment:
1. exact comparison of newSpace and targetSize or EPSILON?
2. --resizables may not be called, is this intentional?  would it be better to keep one statement per line for clarity, even if it adds an extra `if`?